File: ArithmeticSlices4.py
class Solution:
    def numberOfArithmeticSlices(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        def slices(i):
            if i < 2:
                return 0
            ap = 0
            nonlocal result
            if A[i] - A[i - 1] == A[i - 1] - A[i - 2]:
                ap = 1 + slices(i - 1)
                # result[0] += ap
                result += ap
            else:
                slices(i - 1)
            return ap
        # result = [0]
        result = 0
        slices(len(A) - 1)
        # return result[0]
        return result


if __name__ == '__main__':
    A = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.numberOfArithmeticSlices(A))

File: ArithmeticSlices4.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& A) {
    	int n = A.size();
    	if(n < 3) return 0;
    	vector<int> dp(n, 0);
    	int result = 0;
    	for(int i = 2; i < n; i++) {
    		if(A[i] - A[i - 1] == A[i - 1] - A[i - 2]) {
    			dp[i] = dp[i - 1] + 1;
    			result += dp[i];
			}
		}
		
		return result;
    }
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> A;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		A.push_back(num);
	}
	
	Solution *solution = new Solution();
	cout<<solution->numberOfArithmeticSlices(A);
	
	return 0;
	
}
